testplan/common/utils/monitor/__init__.py

```diff
@@ -0,0 +1,269 @@
+"""Resource monitor utilities."""
+
+import os
+import uuid
+import time
+import json
+import socket
+import logging
+import threading
+from testplan.common.utils.path import makedirs
+from .collectors import get_collector
+
+
+class EventMonitor(object):
+    """
+    Event monitor will record event start and stop times(seconds)
+    """
+    def __init__(self, parent=None):
+        self.logger = logging.getLogger('Monitor')
+        self.parent = parent
+        self.hostname = socket.getfqdn()
+        self.events_metadata = {}
+        self.events_data = {}
+        self.lock = threading.Lock()
+        self.log_directory = None
+        self.uid = str(uuid.uuid4())
+
+    def attach(self, event_monitor):
+        """
+        Merge the record event in two EventMonitor
+
+        :param event_monitor:  The EventMonitor will be merged.
+        :type event_monitor: ``EventMonitor``
+        :return:
+        """
+        self.events_metadata.update(event_monitor.events_metadata)
+        self.events_data.update(event_monitor.events_data)
+
+    def setup_logger(self, log_path=None, log_level=None):
+        """
+        Sets up a unique logger for Monitor each time it is called. Should only be called once or logs will be directed
+        to new logger on subsequent call.
+
+        :param log_path: Absolute path to file to save logs to.
+        :type log_path: ``str``
+        :param log_level: Level of logging to save. Currently either ``logging.DEBUG`` if "debug" given or
+                          ``logging.INFO`` if anything else is given.
+        :type log_level: ``str``
+
+        :return: ``None``
+        :rtype: ``NoneType``
+        """
+        self.logger = logging.getLogger(str(uuid.uuid4()))
+        if log_level == 'debug':
+            log_level = logging.DEBUG
+        else:
+            log_level = logging.INFO
+        if log_path:
+            directory = log_path.rsplit(os.sep, 1)[0]
+            self.log_directory = directory
+            makedirs(directory)
+            file_handler = logging.FileHandler(filename=log_path)
+        else:
+            file_handler = logging.NullHandler()  # pylint: disable=bad-option-value
+        formatter = logging.Formatter('%(asctime)s-Monitor-%(levelname)-8s:  %(message)s')
+        file_handler.setFormatter(formatter)
+        file_handler.setLevel(log_level)
+        self.logger.setLevel(log_level)
+        self.logger.addHandler(file_handler)
+
+    def _record_event(self, event_uuid, event):
+        if event_uuid not in self.events_data:
+            self.events_data[event_uuid] = []
+        self.events_data[event_uuid].append({
+            'event': event,
+            'time': time.time(),
+        })
+
+    def started(self, event_type, metadata=None):
```

Comment:
Consider rename started -> start_event? otherwise it is confusing when resource_monitor has a start() method that does totally different thing.

same for stopped

Reply:
event_started()